### DISCUSSION TITLE
Table SelectFilter through 2 relationships?

### DISCUSSION THREAD

**Discussion by yaplej** [filamentphp/filament#2587] [Help] [answered]
Is it possible to filter a Table through two relationships?  I was able to get the data to display on a table but am not figuring out how to get a SelectFilter to work like that.

I was able to bring in the value but not the relationship name.  So I have 1,2,3 as filter options instead of the actual Manufacturer names.
```
SelectFilter::make('manufacturer')
                    ->relationship('assetmodel', 'manufacturer_id')
                    ->label('Manufacturer'),
```
https://filamentphp.com/docs/2.x/tables/filters#relationship-select-filters

**Comment by Z3d0X** [comment, marked as answer]
There is a `->query()` method you can use to customize the query

Something along the lines of this: 
```php
SelectFilter::make('manufacturer')
    ->label('Manufacturer')
    ->options(Manufacturer::pluck('name', 'id'))
    ->query(function (Builder $query, array $data) {
        if (blank($data['value'])) {
            return $query;
        }

        return $query->whereHas('assetmodel', function (Builder $query) use ($data) {
            return $query->where('manufacturer_id', $data['value']);
        });
    }),
 ```

**Reply by yaplej** [reply]
Well that did populate the Filter with the correct data but it does not query the table correctly.

Trying to filter by the Manufacture Name now returns `Object of type Illuminate\Database\Eloquent\Builder is not callable`.

```
                SelectFilter::make('manufacturer')
                    //->relationship('assetmodel', 'manufacturer_id')
                    ->label('Manufacturer')
                    ->options(Company::whereNull('company_id')->pluck('name', 'id'))
                    ->query(function (Builder $query, array $data) {
                        if (blank($data['value'])) {
                            return $query;
                        }
                
                        return $query->whereHas('assetmodel', function (Builder $query) use ($data) {
                            return $query->where('assetmodel.manufacturer.name', $data['value'])();
                        });
                    }),
```

I have tried a few variation for `return $query->where('assetmodel.manufacturer.name', $data['value'])();` but have not been able to get it working correctly.

My table column looks like this.
```
Tables\Columns\TextColumn::make('assetmodel.manufacturer.name')
                    ->label('Manufacturer')
                    ->sortable()
                    ->searchable(),
```

Somehow I would need to join the asset, assetmodel and manufacturer relationships together to get the query to work?

**Reply by Z3d0X** [reply]
> return $query->where('assetmodel.manufacturer.name', $data['value'])();

Remove `()` at the end

**Reply by yaplej** [reply]
That resolved the not callable problem but I am still clearly missing something.

```
SQLSTATE[HY000]: General error: 1 no such column: assetmodel.manufacture.id
select
  count(*) as aggregate
from
  "assets"
where
  (
    exists (
      select
        *
      from
        "asset_models"
      where
        "assets"."assetmodel_id" = "asset_models"."id"
        and "assetmodel"."manufacture"."id" = 2
    )
  )
```

**Reply by Z3d0X** [reply]
>```php
>return $query->whereHas('assetmodel', function (Builder $query) use ($data) {
>    return $query->where('manufacturer_id', $data['value']); //manufacturer_id on asset_models table
>});
> ```

Please see my first reply your are still doing your query wrong. `->where('assetmodel.manufacturer.name'` is not correct
I suggest you take a look at laravel docs https://laravel.com/docs/9.x/eloquent-relationships#querying-relationship-existence

**Reply by yaplej** [reply]
Works perfectly.  Not sure why I was having so much trouble with it at first.
